user-data-store-server/src/main/java/com/wultra/security/userdatastore/service/AttachmentService.java

```diff
@@ -0,0 +1,131 @@
+import com.wultra.security.userdatastore.client.model.request.AttachmentCreateRequest;
+import com.wultra.security.userdatastore.client.model.response.AttachmentCreateResponse;
+import com.wultra.security.userdatastore.client.model.response.AttachmentResponse;
+import com.wultra.security.userdatastore.converter.AttachmentConverter;
+import com.wultra.security.userdatastore.model.entity.AttachmentEntity;
+import com.wultra.security.userdatastore.model.entity.DocumentEntity;
+import com.wultra.security.userdatastore.model.error.ResourceNotFoundException;
+import com.wultra.security.userdatastore.model.repository.AttachmentRepository;
+import com.wultra.security.userdatastore.model.repository.DocumentRepository;
+import lombok.AllArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.time.LocalDateTime;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.UUID;
+
+/**
+ * Service for document attachments.
+ *
+ * @author Roman Strobl, roman.strobl@wultra.com
+ */
+@Service
+@Slf4j
+@AllArgsConstructor
+public class AttachmentService {
+
+    private final AttachmentRepository attachmentRepository;
+    private final DocumentRepository documentRepository;
+    private final Audit audit;
+    private final EncryptionService encryptionService;
+    private final AttachmentConverter attachmentConverter;
+
+    @Transactional
+    public AttachmentCreateResponse createAttachment(final AttachmentCreateRequest request) {
+        final String userId = request.userId();
+        final String documentId = request.documentId();
+        final Optional<DocumentEntity> documentEntityOptional = documentRepository.findById(documentId);
+        if (documentEntityOptional.isEmpty()) {
+            throw new ResourceNotFoundException("Document not found, ID: '%s'".formatted(documentId));
+        }
+        final DocumentEntity documentEntity = documentEntityOptional.get();
+        if (!documentEntity.getUserId().equals(userId)) {
+            throw new ResourceNotFoundException("User reference not valid, ID: '%s'".formatted(userId));
+        }
+        final AttachmentEntity attachmentEntity = new AttachmentEntity();
+        attachmentEntity.setId(UUID.randomUUID().toString());
+        attachmentEntity.setDocument(documentEntity);
+        attachmentEntity.setUserId(userId);
+        attachmentEntity.setAttachmentType(request.attachmentType());
+        attachmentEntity.setTimestampCreated(LocalDateTime.now());
+        encryptionService.encryptAttachment(attachmentEntity, request.attachmentData());
+
+        attachmentRepository.save(attachmentEntity);
+
+        return new AttachmentCreateResponse(attachmentEntity.getId(), documentEntity.getId());
+    }
+
+    @Transactional(readOnly = true)
```

Comment:
Is it read only when it is a writing audit?

Reply:
This should be fine, the audit record is written into a `BlockingQueue` which is flushed asynchronously using a scheduler.

Reply:
👍 Very well spotted! A rare possibility, if the queue is out of capacity. It will be fixed in https://github.com/wultra/lime-java-core/issues/301